### PR TITLE
Add `AccessFilters` utilities

### PR DESCRIPTION
**Pull request by WinstonHartnett**
[`AccessFilters`](https://github.com/WinstonHartnett/bevy/blob/8791f0bf4c0cc2a1b6403398f9c77eb3764e99d8/crates/bevy_ecs/src/query/access.rs#L446) tracks the `With` and `Without` constraints for a `FilteredAccess` set, but is currently private without any way to extend or view underlying filters. This pull just makes `AccessFilters` public, and adds a few utility functions:

- Im/mutable field accessors for `FilteredAccess`
- Functions to add/get `With` and `Without` constraints for `AccessFilters`
- Function to perform in-place union w/ another `AccessFilters`

And updates a few lines to make use of those utilities.

### CODE REVIEW

File: crates/bevy_ecs/src/query/access.rs [hunk 1]
```diff
@@ -303,6 +303,28 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
...
+    }
+
+    /// Returns a mutable reference to the underlying filter sets, collected into disjunctive normal form.
+    ///
+    /// Filter sets are stored in an "ORs of ANDs" format, meaning that a filter `(With<A>, Or<(With<B>, Without<C>)>)`
+    /// expands into `Or<((With<A>, With<B>), (With<A>, Without<C>))>`. All [`AccessFilters`] in the filter sets vector
+    /// are implicitly joined with an `Or` clause, and each `AccessFilters` stores a set of `With` and `Without`
+    /// constraints.
+    #[inline]
+    pub fn filter_sets_mut(&mut self) -> &mut Vec<AccessFilters<T>> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:324]: Can this method be exposed as a safe Rust function?

File: crates/bevy_ecs/src/query/access.rs [hunk 2]
```diff
@@ -452,6 +469,53 @@ impl<T: SparseSetIndex> Default for AccessFilters<T> {
...
+    /// Perform an in-place union with `other`'s corresponding `With` and `Without` sets.
+    #[inline]
+    pub fn union_with(&mut self, other: &Self) {
+        self.with.union_with(&other.with);
+        self.without.union_with(&other.without);
+    }
+
+    /// Returns whether a `With` constraint exists for `index`.
+    #[inline]
+    pub fn get_with(&self, index: T) -> bool {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/access.rs:497]: These methods are probably clearer as "contains_with" etc.

**Comment by alice-i-cecile** [comment]
Initial impressions:

1. Really like the new helper methods.
2. Confused about why you'd like this to be `pub`: can you get into your use case?
3. Very nervous about exposing mutable access, especially in safe code.

**Comment by alice-i-cecile** [comment]
@james-j-obrien, can you take a look at this? Seems like it might be related to https://github.com/bevyengine/bevy/pull/9774.

File: crates/bevy_ecs/src/query/access.rs (hunk 1 above)
Comment (james-j-obrien) [reply crates/bevy_ecs/src/query/access.rs:324]: I am concerned that this would allow an implementor of WorldQuery to hit UB in the case that:
1. The first term of a query is &T and so adds With<T> to the filters
2. The second term (since it now has mutable access to the filter sets in update_component_access) uses the new methods to remove the With<T> requirement.
3. You iterate the query and match an archetype without T hitting UB
Technically implementing WorldQuery is unsafe so allowing this wouldn't be breaking any guarantees, but I'm not sure what the use cases for direct access to the filters are that would make it worth exposing it.

**Comment by james-j-obrien** [comment]
I like the new helper methods, it does improve readability slightly in the places they are used. A bit concerned about the mutable access to the filter sets, while it's probably "safe" it seems like an easy way to break things and I'm not sure when you would want to be going in and editing the filters rather than just adding to them.

File: crates/bevy_ecs/src/query/access.rs (hunk 1 above)
Comment (WinstonHartnett) [reply crates/bevy_ecs/src/query/access.rs:324]: Yeah, that'd be UB. A `FilteredAccess:iter_filter_sets_mut` method would be safe and still fit my use case.

Comment (james-j-obrien) [reply crates/bevy_ecs/src/query/access.rs]: This is safer but you can still edit the filters to remove requirements. What use case are you servicing here?

Comment (WinstonHartnett) [reply crates/bevy_ecs/src/query/access.rs]: Actually, in retrospect, even a mutable iterator could easily allow UB (i.e. by accidentally narrowing a filter with `AccessFilters::add_with` or loosening it with `add_without`, instead of using the safe `and_with` and `and_without` methods)...

**Comment by alice-i-cecile** [comment]
IMO if you split out the mutable accessors into their own PR, we can merge the rest of this dramatically quicker. I still don't understand the use case, and without that I'm opposed to merging them.

**Comment by WinstonHartnett** [comment]
> 3. Very nervous about exposing mutable access, especially in safe code.

- `iter_filter_sets_mut` or `filter_sets_mut` is probably too hard to use safely, so I'll remove it.
- I'd prefer that `AccessFilters` kept its mutable methods, as they're still useful for modifying filter constraint metadata separately from `FilteredAccess`, but I'll remove them from this PR (`AccessFilters` is trivial to reimplement).

> IMO if you split out the mutable accessors into their own PR, we can merge the rest of this dramatically quicker. I still don't understand the use case, and without that I'm opposed to merging them.

Sorry for the slow reply (I was posting between classes).

Long story short, I'm trying to make a `SystemParam` with statically-known accesses and filters (via custom traits), but can't implement it by composing existing `SystemParam`s (`ParamSet`, etc.). I needed a way to create custom `AccessFilters`, get their constraints, apply them to a `FilteredAccess`, and modify each existing filter clause on an ad-hoc basis (hence the mutable and `pub` methods) to make the scheduler happy.

It's clear, however, that the mutable functions are too easy to misuse, so I'll remove them.

**Review by alice-i-cecile — APPROVED** [review]
Okay cool, that's very helpful :) Thanks for the explanation, and sorry for being a bit grumpy.

This looks good to me now: very happy to expose constructors and the utility methods are solid.
